Someone wanted plants?

DisasterMoo opened the pull request:
- Ported 1.12 plants into 1.15.
- Added Flora json reloader, as discussed via discord.
- Loot tables, but still need to change drops to thatch when knife harvested.

Oh right, TODO:
- Flora data placement for all plants, using the spreadsheet somehow.

DisasterMoo committed:
Added all standard, short grass, tall grass, cactus, creeping, epiphyte and hanging plants.

DisasterMoo committed:
Finished adding Plants.

DisasterMoo committed:
Merge branch '1.15.x' of https://github.com/TerraFirmaCraft/TerraFirmaCraft into 1.15.x_dev

DisasterMoo committed:
Added flora feature to world gen:
- Delegates generation to flora types, loaded by json.
- Added plant flora type, which places plants in world.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/plant/TFCBushBlock.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    protected boolean isValidGround(BlockState state, IBlockReader worldIn, BlockPos pos)
+    {
+        return super.isValidGround(state, worldIn, pos) || TFCTags.GRASS.contains(state.getBlock()) || Tags.Blocks.DIRT.contains(state.getBlock());
```
This is a great opportunity to introduce a tag `tfc:bush_plantable_on`.
Vanilla has precedent for this - bamboo does this.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/plant/TFCTallGrassBlock.java:
```diff
@@ -0,0 +1,129 @@
+    public void harvestBlock(World worldIn, PlayerEntity player, BlockPos pos, BlockState state, @Nullable TileEntity te, ItemStack stack)
+    {
+        super.harvestBlock(worldIn, player, pos, Blocks.AIR.getDefaultState(), te, stack);
+    }
```
Why?

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/plant/TFCTallGrassBlock.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public void onBlockHarvested(World worldIn, BlockPos pos, BlockState state, PlayerEntity player)
```
no no no no no this should all be loot table.

DisasterMoo replied:
I've copied vanilla's `DoublePlantBlock#onBlockHarvested`, otherwise, how would a 2 block structure drop a single item?

alcatrazEscapee replied:
Ah, I didn't catch that, my bad.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/plant/PlantBlock.java:
```diff
@@ -0,0 +1,85 @@
+     * Only way to do without having to create a class for each plant with different number of stages
+     * Or ATing into {@link Block}'s stateContainer to make it not final
+     */
+    public abstract Plant getPlant();
```
Having this here basically restricts any possible usage of this class by addons. There's two options:
1) Return an interface here, and implement said interface in `Plant` (basically `IPlant`). Can also make `Plant` just a "fake" enum (e.g. static class with instances)
 - Pros: it works, it's pretty dead simple
 - Cons: it means addons now have to implement this boring interface, which is already perfectly done by an enum. Hello "generic IPlant class"
2) Detach the block from ever holding a reference to the `Plant`, and instead only reference it's *properties*. For instance, add default implementations of `getSpeedFactor`, `getMaxStage`, and `getStage` directly to `PlantBlock`. This is the approach I've taken so far with rock type blocks.
 - Pros: addons can use this directly, just by extending and/or passing in the correct constructor parameters (minimal effort). It also enforces the idea that our `Plant` enum is *only* for block registration, not for querying later.
 - Cons: more work to do.

DisasterMoo committed:
A few changes to generation rules:
- IVeinRule is now IChunkRule (since it can work for everything chunk related in generation).
- Added BiomeRule: Accepts BiomeDictionary tag and biome IDs.
- FloraType generation now accepts IChunkRules for generation.

DisasterMoo committed:
Dumb intellij >:(

DisasterMoo committed:
A bit improvement

DisasterMoo committed:
Updated to keep up with main repo.

DisasterMoo committed:
Cleanup

DisasterMoo committed:
Fixed loot tables.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/TFCBlocks.java:
```diff
@@ -89,7 +93,14 @@
     private static <T extends Block> RegistryObject<T> register(String name, Supplier<T> blockSupplier, Item.Properties blockItemProperties)
     {
         RegistryObject<T> block = BLOCKS.register(name, blockSupplier);
-        TFCItems.ITEMS.register(name, () -> new BlockItem(block.get(), blockItemProperties));
+        TFCItems.ITEMS.register(name, () -> {
+            Block instance = block.get();
+            if (instance instanceof IBlockItemSupplier)
+            {
+                return ((IBlockItemSupplier) instance).apply(blockItemProperties);
+            }
+            return new BlockItem(block.get(), blockItemProperties);
+        });
```
This really grinds my gears, but I can't really articulate why. The main reason is it spreads the registration (e.g. constructor properties) out into every single block, rather then everything being consolidated in one class (e.g. `TFCBlocks`). We do consolidate most item blocks rather than have separate `ItemBlock` registry objects in `TFCItems`, I'm not sure if that's good or bad. But this marker interface I really would rather avoid.

DisasterMoo replied:
Just this one, i'm not sure how else I can delegate the constructor for specific `Blockitem`s when needed besides using this method. If you have something in mind let me know

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/plant/TFCBushBlock.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    protected boolean isValidGround(BlockState state, IBlockReader worldIn, BlockPos pos)
+    {
+        return super.isValidGround(state, worldIn, pos) || TFCTags.Blocks.GRASS.contains(state.getBlock()) || Tags.Blocks.DIRT.contains(state.getBlock());
```
Tag for `bush_plantable_on` which references these two tags.

alcatrazEscapee reviewed (changes requested):
I haven't really tried to comment on the world generation. Given what we now know about 1.16.2 it seems that a lot of the feature style world generation (namely any configurable aspects to it) will need to be heavily rewritten to take advantage of that. However this is more comments on the non-actual-world-gen aspects, because as far as I'm aware those are not getting major changes in the near future.

DisasterMoo commented:
Will get to this soon:tm:

alcatrazEscapee commented:
Just an FYI, I may get to this next, depending on what other 1.16.x things crash and burn in the meantime. But as it stands this PR needs to be very nearly rewritten to catch up with the 1.15 -> 1.16 update, along with mapping changes and world generation rewrites.

DisasterMoo commented:
> Just an FYI, I may get to this next, depending on what other 1.16.x things crash and burn in the meantime. But as it stands this PR needs to be very nearly rewritten to catch up with the 1.15 -> 1.16 update, along with mapping changes and world generation rewrites.

Sure, gonna do something this weekend

DisasterMoo committed:
Merge branch '1.15.x' of https://github.com/TerraFirmaCraft/TerraFirmaCraft into plants

 Conflicts:
	resources/constants.py
	src/main/java/net/dries007/tfc/api/MetalItem.java
	src/main/java/net/dries007/tfc/api/calendar/CalendarEventHandler.java
	src/main/java/net/dries007/tfc/client/ClientEventHandler.java
	src/main/java/net/dries007/tfc/config/ServerConfig.java
	src/main/java/net/dries007/tfc/objects/TFCItemGroup.java
	src/main/java/net/dries007/tfc/objects/blocks/TFCBlocks.java
	src/main/java/net/dries007/tfc/world/biome/BiomeVariantHolder.java
	src/main/java/net/dries007/tfc/world/biome/TFCBiome.java
	src/main/java/net/dries007/tfc/world/feature/TFCFeatures.java
	src/main/resources/assets/tfc/lang/en_us.json
	src/main/resources/assets/tfc/textures/block/plant/porcini/porcini.png
	src/main/resources/assets/tfc/textures/block/plant/rough_horsetail/rough_horsetail.png

DisasterMoo committed:
Updated to keep on track

DisasterMoo committed:
Fixed things for PR

DisasterMoo commented:
Not sure if this will be needed in the future, gonna keep the branch in my repo, otherwise, closing the PR.
